trino/client.py

```diff
@@ -925,3 +977,217 @@ def _parse_retry_after_header(retry_after):
         retry_date = parsedate_to_datetime(retry_after)
         now = datetime.utcnow()
         return (retry_date - now).total_seconds()
+
+
+# Trino Spooled protocol transfer objects
+class _SpooledProtocolResponseTO(TypedDict):
+    encoding: Literal["json", "json+std", "json+lz4"]
+    segments: List[_SegmentTO]
+
+
+class _SegmentMetadataTO(TypedDict):
+    uncompressedSize: str
+
+
+class _SegmentTO(_SegmentMetadataTO):
+    type: Literal["spooled", "inline"]
+    metadata: _SegmentMetadataTO
+
+
+class _SpooledSegmentTO(_SegmentTO):
+    uri: str
+    ackUri: str
+    headers: Dict[str, List[str]]
+
+
+class _InlineSegmentTO(_SegmentTO):
+    data: str
+
+
+class Segment:
+    """
+    Abstract base class representing a segment of data returned by a Trino query.
+
+    Attributes:
+        metadata (property): Metadata associated with the segment.
+        rows (property): Returns the decoded and mapped rows of data.
+    """
+    def __init__(self, segment: _SegmentTO) -> None:
+        self._segment = segment
+        self.metadata = segment["metadata"]
+
+    @property
+    @abstractmethod
+    def rows(self):
+        pass
+
+
+class InlineSegment(Segment):
+    """
+    A subclass of Segment that handles inline data segments. The data is base64 encoded and
+    requires mapping to rows using the provided row_mapper.
+
+    Attributes:
+        rows (property): The rows of data in the segment, decoded and mapped from the base64 encoded data.
+    """
+    def __init__(self, segment: _InlineSegmentTO, row_mapper: RowMapper) -> None:
+        super().__init__(segment)
+        self._segment = cast(_InlineSegmentTO, segment)
+        self._row_mapper = row_mapper
+
+    @property
+    def rows(self) -> List[List[Any]]:
+        inline_segment = self._segment
+        data = inline_segment["data"]
+        decoded_string = base64.b64decode(data)
+        rows = self._row_mapper.map(json.loads(decoded_string))
```

Comment:
we should probably combine _row_mapper with the decoder logic so that it accepts

_row_mapper.map(encoding: str, data: str)

Reply:
The row mapper actually doesn't take a string value but an array of values (each element of the array being a column of the row).

the row mapper only comes into play after decoding the json.